src/amo/components/GetFirefoxBanner/index.js

```diff
@@ -0,0 +1,107 @@
+/* @flow */
+import * as React from 'react';
+import { connect } from 'react-redux';
+import { compose } from 'redux';
+
+import Button from 'amo/components/Button';
+import Notice from 'amo/components/Notice';
+import { DOWNLOAD_FIREFOX_BASE_URL } from 'amo/constants';
+import tracking from 'amo/tracking';
+import { makeQueryStringWithUTM } from 'amo/utils';
+import { isFirefox } from 'amo/utils/compatibility';
+import translate from 'amo/i18n/translate';
+import { replaceStringsWithJSX } from 'amo/i18n/utils';
+import type { UserAgentInfoType } from 'amo/reducers/api';
+import type { AppState } from 'amo/store';
+import type { I18nType } from 'amo/types/i18n';
+
+import './styles.scss';
+
+export const GET_FIREFOX_BANNER_CLICK_ACTION = 'download-firefox-banner-click';
+export const GET_FIREFOX_BANNER_CLICK_CATEGORY = 'AMO Download Firefox';
+export const GET_FIREFOX_BANNER_UTM_CONTENT = 'banner-download-button';
+
+export type Props = {||};
+
+export type DeafultProps = {|
+  _tracking: typeof tracking,
+|};
+
+type PropsFromState = {|
+  userAgentInfo: UserAgentInfoType,
+|};
+
+type InternalProps = {|
+  ...Props,
+  ...DeafultProps,
+  ...PropsFromState,
+  i18n: I18nType,
+|};
+
+export const GetFirefoxBannerBase = ({
+  _tracking = tracking,
+  i18n,
+  userAgentInfo,
+}: InternalProps): null | React.Node => {
+  const onButtonClick = () => {
+    _tracking.sendEvent({
+      action: GET_FIREFOX_BANNER_CLICK_ACTION,
+      category: GET_FIREFOX_BANNER_CLICK_CATEGORY,
+    });
+  };
+
+  if (isFirefox({ userAgentInfo })) {
+    return null;
+  }
+
+  const bannerContent = replaceStringsWithJSX({
+    text: i18n.gettext(
+      `To use these add-ons, you'll need to %(linkStart)sdownload Firefox%(linkEnd)s`,
+    ),
+    replacements: [
+      [
+        'linkStart',
+        'linkEnd',
+        (text) => (
+          <>
+            <br />
+            <Button
+              buttonType="none"
+              className="GetFirefoxBanner-button"
+              href={`${DOWNLOAD_FIREFOX_BASE_URL}${makeQueryStringWithUTM({
+                utm_content: GET_FIREFOX_BANNER_UTM_CONTENT,
+              })}`}
+              onClick={onButtonClick}
+            >
+              {text}
+            </Button>
+          </>
```

Comment:
out of interest, why this complicated extraction for the content of the banner?    Can't you just include this directly in the return below - it's only `{text}` that has localised content.

Reply:
I might not be understanding your question, so forgive me if I'm providing unnecessary information. The full sentence is "To use these add-ons, you'll need to download Firefox", so we want that to be a translatable string - we don't want to break it apart into two pieces. We also need the "download Firefox" to be surrounded with a link, and that link needs an `onClick` handler, so it's not as straightforward as just substituting in `<a href="...">` for `linkStart`. We want to be able to substitute a `Button` component in for `linkStart` and that's what this `replaceStringsWithJSX` function allows us to do.

Does that answer your question?

Reply:
> Does that answer your question?

yes and no 😄 

I get _why_ we're doing it, but I don't really follow how the value of the `text` arg is used in the following replacements arg.  Like, from other frameworks/languages, I'd expect a string with two placeholders in it to have two values to substitute into it, but here we have a list with a list inside, with the first two items of the list just being the names of the placeholders and the final a function that takes (I'm assuming, a different?) `text` arg and includes that arg within a chunk of jsx.  The end result is something... that I assume looks more like the return from `replacements[0][2]` than the original `text`.  

But it's fine, it's not worth us getting into the implementation details of a utility function, I was just asking because it just looked wildly different than I would have expected.